routing/missioncontrol_store.go

```diff
@@ -283,13 +297,17 @@ func (b *missionControlStore) run() {
 	b.wg.Add(1)
 
 	go func() {
-		ticker := time.NewTicker(b.flushInterval)
-		defer ticker.Stop()
+		// TimerChan is non-null when there are items to be stored.
+		var timerChan <-chan time.Time
 		defer b.wg.Done()
 
 		for {
 			select {
-			case <-ticker.C:
+			case <-b.queueChan:
+				timerChan = time.After(b.flushInterval)
+
+			case <-timerChan:
+				timerChan = nil
```

Comment:
The ticker waking up _every_ second is the annoying part on mobile. This prevents the runtime from idling this goroutine, increasing CPU usage. Granted, it's not by much, but afaict, based on some profiling, this is the only tight loop that does not depend on a long (minute+) timer or on an external event (channel update, gossip msg, p2p block, etc) happening, so it seemed it could easily be improved with this fix.

Reply:
Android has tight restrictions for background operation (no more than 6s of cpu time per 5min interval before an app is killed), so idling as much as possible is desirable.

Reply:
Ah, I see. I think you should add that to the commit message, otherwise the next reviewer might ask the same question.

But, this pattern with the channel is very weird. Can you implement this using `sync.Cond` please? Then we don't need to drain any channels or anything like that at the end.

Reply:
I'm not sure how to use `Cond` in this situation, because Cond doesn't allow selecting on it (which is needed here due to `b.done`). 

I guess I could use a timer instead of chans directly here or switch back to the ticker, but stop it if there's no work if that would make more sense to you (though I find this particular pattern of a nil chan shorter and clearer when the timeouts are small enough that you don't really need to stop the timer).

Reply:
@matheusd  - I think you can do it as follows:

1) have these members in the `missionControlStore`

```
	queue     *list.List
	queueMx   sync.Mutex
	queueCond *sync.Cond
```

2) initialise `queueCond`:
```
mc.queueCond = sync.NewCond(&mc.queueMx)
```

3) replace all calls to `queueMx` with `queueCond.L`

4) update `AddResult`:

```
func (b *missionControlStore) AddResult(rp *paymentResult) {
	b.queueCond.L.Lock()
	b.queue.PushBack(rp)
	b.queueCond.L.Unlock()

	b.queueCond.Signal()
}
```

5) update `run`: 

```
func (b *missionControlStore) run() {
	b.wg.Add(1)

	go func() {
		defer b.wg.Done()

		timer := time.NewTimer(b.flushInterval)
		timer.Stop()

		for {
			// Wait for the queue to not be empty.
			b.queueCond.L.Lock()
			for b.queue.Front() == nil {
				b.queueCond.Wait()

				select {
				case <-b.done:
					b.queueCond.L.Unlock()

					return
				default:
				}
			}
			b.queueCond.L.Unlock()

			// Restart the timer.
			timer.Reset(b.flushInterval)

			select {
			case <-timer.C:
				if err := b.storeResults(); err != nil {
					log.Errorf("Failed to update mission "+
						"control store: %v", err)
				}

			case <-b.done:
				return
			}
		}
	}()
}
```

6) adjust `stop`:

```
// stop stops the store ticker goroutine.
func (b *missionControlStore) stop() {
	close(b.done)

	b.queueCond.Signal()

	b.wg.Wait()
}
```

Reply:
Personally, I find `Cond` and `Timer` harder to reason about then just standard channels (but that's really on me).

In any case, I updated the code with your suggested implementation. Thank you for this!